nntrainer/dataset/data_producers.h

```diff
@@ -43,12 +44,31 @@ class DataProducer {
 
   /**
    * @brief create an iteration
+   * @todo rename this to BatchGenerator
    * @return Iteration iteration, if std::get<0>(retval) == true means end of
    * iteration, at the end of the iteration, it's responsibility of @a this to
    * shuffle.
    */
   using Generator = std::function<Iteration(void)>;
 
+  /**
+   * @brief Sample represents a view of single element which can be fed to the
+   * model. It is the smallest unit to produce a data
+   * @return std::get<0>(Sample) denotes inputs
+   * @return std::get<1>(Sample) denotes labels
+   */
+  using Sample = std::tuple<std::vector<Tensor *>, std::vector<Tensor *>>;
+
+  /**
+   * @brief generator callable type which will fill a sample
+   * @todo rename this to Generator.
+   * @param[out] Sample @see DataProducer::Sample for detail, inside this Sample
+   * must contain allocated tensor before fed to this function.
+   * @return bool true if this is the last sample.
+   *
+   */
+  using Generator_sample = std::function<bool(Sample &)>;
```

Comment:
consider directly using `std::vector<Tensor *>, std::vector<Tensor *>` or update the `size()` to have a similar tuple based interface.

Reply:
Sample will be a class onward.
I guess we can use ConnectionSpecs  (or SampleSpec if we need) later on for the size query.

Reply:
But I think it's better to pass std::vector<Tensor *>, std::vector<Tensor *> here. I'll change.

Reply:
ok.

Reply:
Done